cas-server-support-oauth/src/main/java/org/jasig/cas/support/oauth/web/OAuth20RevokeClientTokensController.java

```diff
@@ -68,11 +72,32 @@ protected ModelAndView handleRequestInternal(final HttpServletRequest request, f
             return OAuthUtils.writeJsonError(response, OAuthConstants.INVALID_REQUEST, e.getMessage(), HttpStatus.SC_BAD_REQUEST);
         }
 
-        if (!centralOAuthService.revokeClientTokens(clientId, clientSecret)) {
+        // verify that the client service has a valid client_id and client_secret
+        final OAuthRegisteredService service = centralOAuthService.getRegisteredService(clientId);
+        if (service == null || !service.getClientSecret().equals(clientSecret)) {
             LOGGER.error("Could not revoke client tokens, mismatched client id or client secret");
-            return OAuthUtils.writeJsonError(response, OAuthConstants.INVALID_REQUEST,
+            return OAuthUtils.writeJsonError(
+                    response,
+                    OAuthConstants.INVALID_REQUEST,
                     OAuthConstants.INVALID_CLIENT_ID_OR_SECRET_DESCRIPTION,
-                    HttpStatus.SC_BAD_REQUEST);
+                    HttpStatus.SC_BAD_REQUEST
+            );
+        }
+
+        // the MAX_NUM_OF_ATTEMPTS prevents potential deadlock
+        for (int i = 0; i <= MAX_NUM_OF_ATTEMPTS; i++) {
```

Comment:
Set a `MAX_NUM_OF_ATTEMPTS` in case some token and related tickets are locked. Raises `SC_INTERNAL_SERVER_ERROR` if the cap is reached.